docs/client.rst

```diff
@@ -0,0 +1,78 @@
+.. _client:
+
+The client object
+=================
+
+Storing and restoring client state
+----------------------------------
+
+The :class:`~fints.client.FinTS3Client` object keeps some internal state that's beneficial to keep
+across invocations. This includes
+
+ * A system identifier that uniquely identifies this particular FinTS endpoint
+ * The Bank Parameter Data (BPD) with information about the bank and its advertised capabilities
+ * The User Parameter Data (UPD) with information about the user account and allowed actions
+
+.. autoclass:: fints.client.FinTS3Client
+   :members: __init__, get_data, set_data
+   :noindex:
+   :undoc-members:
+
+Using the :func:`~fints.client.FinTS3Client.get_data`/:func:`~fints.client.FinTS3Client.set_data`
+facility is purely optional for reading operations, but may speed up the process because the BPD/UPD
+can be cached and need not be transmitted again.
+
+It may be required to use the facility for transaction operations if both parts of a two-step transaction
+cannot be completed with the same :class:`~fints.client.FinTS3Client` object.
+
+The :func:`~fints.client.FinTS3Client.get_data` parameter `include_private` (defaults to `False`) enables
+including the User Parameter Data in the datablob. Set this to `True` if you can sufficiently ensure the
+privacy of the returned datablob (mostly: user name and account numbers).
+
+If your system manages multiple users/identity contexts, you SHOULD keep distinct datablobs per
+user or context.
+
+You SHOULD NOT call any other methods on the :class:`~fints.client.FinTS3Client` object
+after calling :func:`~fints.client.FinTS3Client.get_data`.
```

Comment:
Hmm, the naming idea is a good one, that captures the spirit better. Reason: I wanted to keep the simplicity of not *having* to do a deconstruct/teardown or context in most cases (for the simple "getting started" in the docs, where it does a `client = FinTS3Client(...); client.do_foo()`), but the *ability* to save state if feasible. (I my use case, I'm wrapping this in a context manager that handles storing/restoring from database: https://github.com/henryk/byro-fints/blob/c71c6d576b6b1c62037c760ade9db4f084c41ae7/byro_fints/views.py#L138)

As to what would happen: Right now, in PIN/TAN, nothing bad, that's why it's a "SHOULD NOT". It's simply caching data that won't need to be retrieved again. But: I'm already thinking the API in terms of potential other mechanisms. For HBCI (with RSA key file) I think we will need to manage a signature sequence number on our end, so that would need to be stored somewhere.

That's why it was originally called get_data/set_data: Because the Client object may have a need to store data between instantiations, but I won't mandate where it's stored or how. That shoud be up to the library user.
(If you look at HBCI4Java they solve that in a different way, IIRC: They have something called passport files where state is stored, and the library updates them on its own.)

Reply:
> Hmm, the naming idea is a good one, that captures the spirit better. Reason: I wanted to keep the simplicity of not having to do a deconstruct/teardown or context in most cases (for the simple "getting started" in the docs, where it does a client = FinTS3Client(...); client.do_foo()), but the ability to save state if feasible. (I my use case, I'm wrapping this in a context manager that handles storing/restoring from database: henryk/byro-fints:byro_fints/views.py@c71c6d5#L138)

Yes, I've seen that. Okay, then I maybe rename it to ``deconstruct()``? I'm open for other proposals as well, but that's what Django uses in some places for things like this.

> As to what would happen: Right now, in PIN/TAN, nothing bad, that's why it's a "SHOULD NOT". It's simply caching data that won't need to be retrieved again. But: I'm already thinking the API in terms of potential other mechanisms. For HBCI (with RSA key file) I think we will need to manage a signature sequence number on our end, so that would need to be stored somewhere.

I've noticed that you already kept HBCI-auth in mind. I don't feel it is hugely important, since Commerzbank is the *only* known bank out there supporting key file auth, but not PIN/TAN, and my feeling is that key file auth will die with PSD2, but as far as I know that's not yet for sure.

> That's why it was originally called get_data/set_data: Because the Client object may have a need to store data between instantiations, but I won't mandate where it's stored or how. That shoud be up to the library user.
> (If you look at HBCI4Java they solve that in a different way, IIRC: They have something called passport files where state is stored, and the library updates them on its own.)

I agree, this should be left to the library user.